packages/contracts-periphery/contracts/universal/op-nft/Optimist.sol

```diff
@@ -0,0 +1,160 @@
+        require(isWhitelisted(_recipient), "Optimist: address is not whitelisted");
+        _safeMint(_recipient, tokenIdOfOwner(_recipient));
+    }
+
+    /**
+     * @notice  Returns decimal tokenid for a given address
+     * @return  uint256 decimal tokenId
+     */
+    function baseURI() public view returns (string memory) {
+        return
+            string(
+                abi.encodePacked(
+                    ATTESTATION_STATION.attestations(
+                        ATTESTOR,
+                        address(this),
+                        bytes32("optimist.base-uri")
+                    )
+                )
+            );
+    }
+
+    /**
+     * @notice Returns the URI for the token metadata.
+     * @dev The token URI will be stored at baseURI + '/' + tokenId + .json
+     * @param tokenId The token ID to query.
+     * @return The URI for the given token ID.
+     */
+    function tokenURI(uint256 tokenId) public view virtual override returns (string memory) {
+        return
+            string(
+                abi.encodePacked(
+                    baseURI(),
+                    "/",
+                    // convert tokenId to hex string formatted like an address (20)
+                    Strings.toHexString(tokenId, 20),
+                    ".json"
+                )
+            );
+    }
+
+    /**
+     * @notice  Returns whether an address is whitelisted
+     * @dev     The whitelist is an attestation by the admin of this contract
+     * @return  boolean  Whether the address is whitelisted
+     */
+    function isWhitelisted(address _recipient) public view returns (bool) {
+        return
+            ATTESTATION_STATION
+                .attestations(ATTESTOR, _recipient, bytes32("optimist.can-mint"))
+                .length > 0;
+    }
+
+    /**
+     * @notice  Returns decimal tokenid for a given address
+     * @return  uint256 decimal tokenId
+     */
+    function tokenIdOfOwner(address _owner) public pure returns (uint256) {
+        return uint256(uint160(_owner));
+    }
+
+    /**
+     * @notice  (Internal) Soulbound
+     * @dev     Override internal function to prevent transfers of the Optimist token.
+     * @param   _from  The address of the token sender.
+     * @param   _to    The address of the token recipient.
+     */
+    function _beforeTokenTransfer(
+        address _from,
+        address _to,
+        uint256
+    ) internal virtual override {
+        require(_from == address(0) || _to == address(0), "Optimist: soul bound token");
+    }
+
+    /**
+     * @notice  Soulbound
+     * @dev     Override internal function to prevent transfers of the Optimist token.
+     */
+    function approve(address, uint256) public pure override {
+        revert("Optimist: soul bound token");
+    }
+
+    /**
+     * @notice  Soulbound
+     * @dev     Override internal function to prevent transfers of the Optimist token.
+     */
+    function _setApprovalForAll(
```

Comment:
Nit, I would do this for the public function and not the internal function just like you did for approve